Feat: display preview to AI fix

MichaelAquilina opened the pull request:
### Description

Add a new display preview when generating AI Fixes instead of them being applied to the user's code immediately.

The changes can be applied by the user with the "Apply Fix" button if they are happy with the changes being suggested.

If clicked, the updated code will be temporarily highlighted until either:
- 30 seconds have passed
- the user saves the file
- the user changes some text in the file

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] CHANGELOG.md updated
- [ ] README.md updated, if user-facing

### Screenshots / GIFs

![image](https://github.com/snyk/vscode-extension/assets/2650040/b5e1e61b-113d-40e4-a037-e8ee65384ecd)

[Screencast from 2024-04-05 10-13-39.webm](https://github.com/snyk/vscode-extension/assets/2650040/39b63834-148c-4353-8811-18af145b4f19)

[Screencast from 2024-04-17 16-47-21.webm](https://github.com/snyk/vscode-extension/assets/2650040/0efb3eab-1e27-41f1-9b15-670a2455e921)

MichaelAquilina commented on src/snyk/common/views/issueTreeProvider.ts:
```diff
@@ -35,6 +38,14 @@ export abstract class ProductIssueTreeProvider<T> extends AnalysisTreeNodeProvid
     );
   }
 
+  getAIFix(issue: Issue<T>): boolean {
```
do we want to keep this generic or enforce that it is only Code?

Does this make sense to be inside the abstract class? Why not keep this inside the Code implementation?

MichaelAquilina commented on src/snyk/common/views/issueTreeProvider.ts:
```diff
@@ -86,20 +97,31 @@ export abstract class ProductIssueTreeProvider<T> extends AnalysisTreeNodeProvid
+          text: (this as any).getAIFixableIssuesText(AIFixVulnCount),
+        }),
+      );
+    }
```
should be in the code implementation

MichaelAquilina commented on src/snyk/common/views/issueTreeProvider.ts:
```diff
@@ -86,20 +97,31 @@ export abstract class ProductIssueTreeProvider<T> extends AnalysisTreeNodeProvid
+      );
+    }
+
+    // this.getDurationTreeNode(),
```
do we want to remove this?

MichaelAquilina commented on src/snyk/common/views/issueTreeProvider.ts:
```diff
@@ -136,6 +158,10 @@ export abstract class ProductIssueTreeProvider<T> extends AnalysisTreeNodeProvid
 
+          if (this.getAIFix(issue)) {
+            AIFixVulnCoun++;
+          }
```
this should be somewhere else or rename the function to represent it accordingly

the return type also should not be a tuple

MichaelAquilina commented on src/snyk/common/views/issueTreeProvider.ts:
```diff
@@ -209,13 +235,19 @@ export abstract class ProductIssueTreeProvider<T> extends AnalysisTreeNodeProvid
+    return issuesCount && issuesCount > 0
+      ? `⚡️ ${issuesCount} ${issuesCount === 1 ? 'vulnerability' : 'vulnerabilities'} can be fixed by Snyk DeepCode AI`
+      : 'There are no vulnerabilities fixable by Snyk DeepCode AI';
+  }
```
this is also code implementation

MichaelAquilina commented on src/snyk/common/views/issueTreeProvider.ts:
```diff
@@ -209,13 +235,19 @@ export abstract class ProductIssueTreeProvider<T> extends AnalysisTreeNodeProvid
 
+  protected getAIFixableIssuesText(issuesCount: number | null): string {
+    return issuesCount && issuesCount > 0
+      ? `⚡️ ${issuesCount} ${issuesCount === 1 ? 'vulnerability' : 'vulnerabilities'} can be fixed by Snyk DeepCode AI`
```
is there a lodash function to pluralize?

MichaelAquilina commented on src/snyk/common/views/issueTreeProvider.ts:
```diff
@@ -13,6 +13,9 @@ import { Command, Range } from '../../common/vscode/types';
 }
+interface IssueAdditionalData {
+  hasAIFix?: boolean;
+}
```
unused, can be deleted

MichaelAquilina commented on src/snyk/common/views/issueTreeProvider.ts:
```diff
@@ -35,6 +38,14 @@ export abstract class ProductIssueTreeProvider<T> extends AnalysisTreeNodeProvid
+    if (myIssue.additionalData.hasAIFix) {
+      return true;
+    }
+    return false;
```
this seems to be dead code, its re-implemented in the snykCode/issueTreeProvider

MichaelAquilina commented on src/snyk/snykCode/views/issueTreeProvider.ts:
```diff
@@ -33,8 +33,18 @@ export class IssueTreeProvider extends ProductIssueTreeProvider<CodeIssueData> {
+  }
+
+  getAIFix(issue: Issue<CodeIssueData>): boolean {
+    return issue.additionalData.hasAIFix ? true : false;
```
```suggestion
    return issue.additionalData.hasAIFix;
```

MichaelAquilina commented on src/snyk/snykCode/views/securityIssueTreeProvider.ts:
```diff
@@ -39,8 +39,15 @@ export default class CodeSecurityIssueTreeProvider extends IssueTreeProvider {
-    return `Snyk found ${
-      !nIssues ? 'no vulnerabilities! ✅' : `${nIssues} ${nIssues === 1 ? 'vulnerability' : 'vulnerabilities'}`
-    }`;
+    // return `Snyk found ${!nIssues ? 'no vulnerabilities! ✅' : `${nIssues} ${nIssues === 1 ? 'vulnerability' : 'vulnerabilities'}`
```
dead code to be removed

MichaelAquilina commented on src/snyk/snykCode/views/securityIssueTreeProvider.ts:
```diff
@@ -39,8 +39,15 @@ export default class CodeSecurityIssueTreeProvider extends IssueTreeProvider {
+    // }`;
+    let message = "";
+
+    if (nIssues <= 0) {
```
```suggestion
    if (nIssues > 0) {
```

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -1,14 +1,23 @@
-import _ from 'lodash';
+import _, { forEach } from 'lodash';
```
unused forEach

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -152,13 +161,24 @@ export class CodeSuggestionWebviewProvider
+    // get the workspace folders
+    // look at the filepath and identify the folder that contains the filepath
+    let workspaceFolder = '';
+    for (const element of this.workspace.getWorkspaceFolders()) {
```
element -> folder / path

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -152,13 +161,24 @@ export class CodeSuggestionWebviewProvider
       id: issue.id,
       title: issue.title,
-      uri: issue.filePath,
+      uri: issue.filePath.replace(workspaceFolder + '/', ''),
```
we should clean this up

we need this to make the filePath relative, which is needed for the ls

maybe might even make sense to move the logic out?

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -204,6 +224,53 @@ export class CodeSuggestionWebviewProvider
+        case 'applyGitDiff': {
+          const { patch, fileUri } = args;
+
+          const fileContent = readFileSync(fileUri, 'utf8');
```
open question, is it ok to assume utf-8?

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -204,6 +224,53 @@ export class CodeSuggestionWebviewProvider
+
+            const editor = vscode.window.visibleTextEditors.find(
+              editor => editor.document.uri.toString() === `file://${fileUri}`,
+            );
```
should we open the editor if we have not found it here?

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -204,6 +224,53 @@ export class CodeSuggestionWebviewProvider
+              editor => editor.document.uri.toString() === `file://${fileUri}`,
+            );
+
+            if (!editor) return console.error('No editor found');
```
better error handling

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -204,6 +224,53 @@ export class CodeSuggestionWebviewProvider
+
+            if (!editor) return console.error('No editor found');
+            const editorEndLine = editor.document.lineCount;
+            edit.replace(
```
open question, maybe there is a better API replacing a whole file?

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -204,6 +224,53 @@ export class CodeSuggestionWebviewProvider
+            const success = await vscode.workspace.applyEdit(edit);
+            if (success) {
+              this.highlightAddedCode(fileUri, patch);
+            } else console.error('No good');
```
better error handling

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -204,6 +224,53 @@ export class CodeSuggestionWebviewProvider
+            } else console.error('No good');
+          }
+
+          // void this.panel?.webview.postMessage({ type: 'asd', args: {} });
```
dead code

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -213,6 +280,78 @@ export class CodeSuggestionWebviewProvider
     }
   }
 
+  private highlightAddedCode(fileUri: string, patch: string) {
```
open question, should this be hoisted out of the class into a function?

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -213,6 +280,78 @@ export class CodeSuggestionWebviewProvider
 
+  private highlightAddedCode(fileUri: string, patch: string) {
+    const highlightDecoration = vscode.window.createTextEditorDecorationType({
+      backgroundColor: 'rgba(0,255,0,0.3)',
```
this needs to be theme compatible

MichaelAquilina replied:
look into if something here makes sense to use: https://code.visualstudio.com/api/references/theme-color

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -404,46 +524,101 @@ declare const acquireVsCodeApi: any;
-    suggestionDetailsElem.classList.add('collapsed');
-    readMoreBtnElem.classList.remove('hidden');
-    suggestionDetailsContentElem.classList.remove('hidden');
+  function toggleLoading(toggle: string) {
```
same as above

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -404,46 +524,101 @@ declare const acquireVsCodeApi: any;
     }
   }
 
+  function generateAIfix() {
```
generateAIFixHandler?

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -489,6 +667,23 @@ declare const acquireVsCodeApi: any;
+          toggleFixes('show');
+          toggleLoading('hide');
+          toggleFixWrapper('hide');
+          const state = vscode.getState();
```
unused?

MichaelAquilina replied:
probably can be removed

MichaelAquilina commented on src/snyk/snykOss/providers/ossVulnerabilityTreeProvider.ts:
```diff
@@ -140,7 +140,7 @@ export default class OssIssueTreeProvider extends ProductIssueTreeProvider<OssIs
     }
 
-    return [nodes, totalVulnCount];
+    return [nodes, totalVulnCount, null];
```
can be simplified

carlos-snyk commented on src/snyk/common/views/issueTreeProvider.ts:
```diff
@@ -220,12 +248,6 @@ export abstract class ProductIssueTreeProvider<T> extends AnalysisTreeNodeProvid
   }
 
-  // todo: Obsolete. Remove after OSS scans migration to LS
-  protected getFilteredIssues(diagnostics: readonly unknown[]): readonly unknown[] {
```
this was not in use

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -63,11 +55,8 @@ export class CodeSuggestionWebviewProvider
-  activate(): void {
-    this.context.addDisposables(
-      this.window.registerWebviewPanelSerializer(SNYK_VIEW_SUGGESTION_CODE, new WebviewPanelSerializer(this)),
-    );
```
@andygongea do you remember why this was removed?

MichaelAquilina replied:
this probably needs to be added back

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -25,24 +31,10 @@ import { getAbsoluteMarkerFilePath } from '../../utils/analysisUtils';
-  markers?: Marker[];
-  cols: Point;
-  rows: Point;
-};
```
moved to a dedicated `types.ts` file along with some other new ones

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -152,58 +141,129 @@ export class CodeSuggestionWebviewProvider
     try {
-      const { type, args } = message;
-      switch (type) {
+      switch (message.type) {
```
all the cases should now be typed

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -152,58 +145,113 @@ export class CodeSuggestionWebviewProvider
+              suggestion.uri,
+              this.openIssueId,
+            );
+            this.logger.info(Object.values(diffs[0].unifiedDiffsPerFile)[0]);
```
remove

MichaelAquilina commented on src/test/unit/snykCode/utils/patchUtils.test.ts:
```diff
@@ -0,0 +1,117 @@
+ print(result)
+
+-result *= generate_number()
++result += generate_number()
```
add actual extra hunk to this test case

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewProvider.ts:
```diff
@@ -213,6 +261,62 @@ export class CodeSuggestionWebviewProvider
+
+    editor.setDecorations(highlightDecoration, decorationOptions);
+
+    const firstLine = decorationOptions[0].range.start.line;
```
handle case where its empty

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -146,17 +265,135 @@ declare const acquireVsCodeApi: any;
+  function applyFix() {
+    if (!suggestion) return;
+    const diffSuggestion = suggestion.diffs[diffSelectedIndex];
+    const filePath = suggestion.folderPath + '/' + suggestion.uri;
```
this might break in windows, needs to be fixed

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -146,17 +265,135 @@ declare const acquireVsCodeApi: any;
+
+    const diffSuggestion = suggestion.diffs[diffSelectedIndex];
+
+    const filePath = suggestion.folderPath + '/' + suggestion.uri;
```
needs to support windows

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -146,17 +267,144 @@ declare const acquireVsCodeApi: any;
+    );
+  }
+
+  function generateDiffHtml(diffData: string): HTMLElement {
```
`patch`

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -146,17 +265,135 @@ declare const acquireVsCodeApi: any;
+        }
+      } else {
+        const lineDiv = document.createElement('div');
+        lineDiv.className = 'example-line';
```
change to something more appropriate like `code-line`

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -306,12 +420,20 @@ declare const acquireVsCodeApi: any;
+            uniqueRows.add(rowStartStr);
+            markLineText += (first ? '' : ', ') + rowStartStr;
+            first = false;
+          }
```
try remove this and see if anything breaks

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -327,6 +586,7 @@ declare const acquireVsCodeApi: any;
     }
 
+    // TODO: should probably fix this to use toggleElement
     moreInfoElem.className = suggestion.leadURL ? 'clickable' : 'clickable hidden';
```
fix these cases

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -404,51 +665,63 @@ declare const acquireVsCodeApi: any;
+  const loadingIndicator = document.querySelector('.sn-loading');
+  const fixWrapper = document.querySelector('.sn-fix-wrapper');
+  const fixesSection = document.querySelector('.sn-ai-fixes');
+  const fixError = document.querySelector('.sn-ai-fix-error');
```
same thing here

MichaelAquilina commented on src/snyk/snykCode/views/suggestion/codeSuggestionWebviewScript.ts:
```diff
@@ -489,6 +762,34 @@ declare const acquireVsCodeApi: any;
+          break;
+        }
+        toggleElement(fixWrapper, 'hide');
+        toggleElement(fixError, 'show');
```
could do with some style tweaks
![image](https://github.com/snyk/vscode-extension/assets/2650040/e0145334-a64b-41ec-9931-86bd10400b8e)

MichaelAquilina replied:
@andygongea any thoughts here?

MichaelAquilina commented on src/test/unit/snykCode/utils/patchUtils.test.ts:
```diff
@@ -0,0 +1,117 @@
+    assert.strictEqual(result[3].range.end.character, 7);
+  });
+
+  test('empty remove', () => {
```
use TDD "should" style

Arvi3d commented:
https://github.com/snyk/vscode-extension/assets/120917/d011ecf1-4775-4ae9-a43f-e471170f5f1b

Arvi3d commented:
https://github.com/snyk/vscode-extension/assets/120917/3b59c42e-b22a-408d-8bb2-1916138d4832

Arvi3d commented:
<img width="1721" alt="SCR-20240408-qwmv" src="https://github.com/snyk/vscode-extension/assets/120917/ab97a164-59c6-4ad4-902f-9f1d8951a3c2">

Arvi3d commented:
Is Fix AI functional now? I can't get any fixes for juice-shop repo.

Arvi3d commented:
I compared it with IntellyJ IDEA for https://github.com/raphabot-snyk/juice-shop/blob/main/lib/insecurity.js#L21
Both VSC and IDEA find vulns. IDEA also successfully fixes it. While VSC with the new changes, always fails.

bastiandoetsch commented:
@Arvi3d, did it show you diffs? VSCode shouldn't interfere in the contents - they are provided by language server, and VSCode only has the interaction and diff application code.

bastiandoetsch commented:
Yeah, doesn't work for me either - there seems to be a bug calling LS.

MichaelAquilina commented:
@Arvi3d @bastiandoetsch I've updated the PR to fix the issues (there where some typos in the ids I introduced)

In terms of the "Fix Analysis" tab being used to show community fixes, it seems from talking to @andygongea this is intended behaviour

andygongea commented:
@MichaelAquilina @Arvi3d 

Regarding the comment about the tab name, the content is split in two tabs - one about the fix (Fix Analysis) the other about the vuln data (Vulnerability Overview).

When we can't fix and issue with AI Fix, the Community Fixes are displayed providing some fix data, hence fitting into the Fix Analysis tab.

Arvi3d commented:
Hi folks. testing it now again.

MichaelAquilina commented:
I found one weird UI bug which I am investigating right now:

- open a fixable issue
- click generate fix
- wait for fix to be generated
- open another fixable issue
- click generate fix
- wait for fix to be generated
- click anywhere on the editor

The "fix wrapper" is now shown incorrectly:

![image](https://github.com/snyk/vscode-extension/assets/2650040/fcdba8f0-72c1-44ab-b22f-64efd0bfa4e1)

Arvi3d commented:
@andygongea, let's we have this example 
<img width="691" alt="SCR-20240409-lvzn" src="https://github.com/snyk/vscode-extension/assets/120917/bdd04f48-e46c-4015-841d-9aa5eaa85d6e">
NoSQL Injection
```
Unsanitized input from the HTTP request body[:13] flows[:13, :13, :13, :15, :15, :15] into findOne[:15], where it is used in an NoSQL query. This may result in an NoSQL Injection vulnerability.
```
This description does not say anything about fixing the issue.
Sometimes, such a description has more fix-like command tone: "do that, change this, etc", but not always.

So, i'd argue, that it look illogical under "Fix analysis" tab.
Should we move it out? Put it before the tabs?

Arvi3d commented:
Ok, this is an example of timeout'ed request for the fix.
<img width="1728" alt="SCR-20240409-lzex" src="https://github.com/snyk/vscode-extension/assets/120917/3dca0366-e7ae-4ac9-996c-c1949b189839">

Arvi3d commented:
some weird behaviour when i click "Apply fix"
https://drive.google.com/file/d/13TOigpKnH_dQ7OqlusFW_GXu2s1jOEUX/view?usp=sharing

Arvi3d reviewed (dismissed):
.

MichaelAquilina commented:
@Arvi3d @andygongea might be worth us three going over what the expected behaviour is because it seems like a lot of what @Arvi3d talks about in his video is intended functionality from what @andygongea explained to me.

I can understand the confusion so might be worth seeing what the best path forward is. Is it ok I find some time to go over these together?

MichaelAquilina commented:
> Ok, this is an example of timeout'ed request for the fix. <img alt="SCR-20240409-lzex" width="1728" src="https://private-user-images.githubusercontent.com/120917/320829496-3dca0366-e7ae-4ac9-996c-c1949b189839.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI3NjY4NjEsIm5iZiI6MTcxMjc2NjU2MSwicGF0aCI6Ii8xMjA5MTcvMzIwODI5NDk2LTNkY2EwMzY2LWU3YWUtNGFjOS05OTZjLWMxOTQ5YjE4OTgzOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxMFQxNjI5MjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jYzM3ZWNjMTlmNjA2ZWIxNDU3MWRiMTU5NzhjZmJiMzBkNGZkMjc2NDI3YjJiOGFlNDhjYTMyYTA0NzI1NDYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.W-6mFkm78s_I6TAAadGd7eA2vT_VgS6CVPRz07MYgXM">

SO I looked into this one and it seems like the errors are displayed when the language server reports an issue. The errors are unrelated to this change so I'm not sure its worth adding more scope to this PR. I'll put it in the [list of planned improvements](https://docs.google.com/spreadsheets/d/1ICan6wma_2Q2810UB4FxpTia2GPjT5fqxjfsbeqNErE/edit#gid=0) though :)

MichaelAquilina commented:
@Arvi3d @bastiandoetsch thanks for your feedback so far!

I've added all the agreed upon changes and it should be ready for another round of review and testing :)

Arvi3d commented:
ok, testing now... thank you Mike

MichaelAquilina commented:
>  CHANGELOG.md updated
> README.md updated, if user-facing

Not sure what the convention for these two are if someone can help :) I'm guessing merging to main wont release a new version immediately so not sure what exact contents to put in `CHANGELOG.md`

For README.md's case, I dont see any case where it needs to be updated - possibly because AI Fix is still an opt in feature? But I could be wrong :)

Arvi3d commented:
IMHO, error handling is a pretty big deal considering the state of Code Fix AI services. 
In that improvements document, it's not even priority#1. From a UX point of view, I disagree with that.

But also, this is a tech debt, so I'm going to leave it for @bastiandoetsch to decide and unblock it from my side.

MichaelAquilina commented:
> IMHO, error handling is a pretty big deal considering the state of Code Fix AI services. In that improvements document, it's not even priority#1. From a UX point of view, I disagree with that.
> 
> But also, this is a tech debt, so I'm going to leave it for @bastiandoetsch to decide and unblock it from my side.

@bastiandoetsch might have to confirm this for me, but from what I can tell - these errors are actually originating from the Snyk language server not visual studio code. So fixing them would involve making separate changes in the language server (which is out of scope of this PR at the very least)

bastiandoetsch commented:
Didn't manage to review it, yet. It's high on my todo list, though!

bastiandoetsch reviewed (commented):
Error handling still seems to need some improvement - I would expect the executeCommand failed message to not be displayed, if the patch retrieval fails. WDYT?

![image](https://github.com/snyk/vscode-extension/assets/20150761/3125d8ca-43da-43b0-8c13-2abee2d1eb45)

MichaelAquilina commented:
> Error handling still seems to need some improvement - I would expect the executeCommand failed message to not be displayed, if the patch retrieval fails (lower right corner). WDYT?

so I thought this was something that would need to change on the language server, but thinking about it, we show the `Oh snap` message so I must have concluded this wrongly 🤔

bastiandoetsch commented:
> > Error handling still seems to need some improvement - I would expect the executeCommand failed message to not be displayed, if the patch retrieval fails (lower right corner). WDYT?
> 
> so I thought this was something that would need to change on the language server, but thinking about it, we show the `Oh snap` message so I must have concluded this wrongly 🤔

No you haven't, the HTTP 500 triggers the language server command handler to popup the dialog. The Oh Snap! message uses a different call/methodology. We can filter it though - I'll take care of that in the getAutofixDiffs command. If we had better HTTP codes on the backend, it would be easier, as we don't have any chance to report real errors, if we filter on the language server side.

MichaelAquilina commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
 
+## [2.4.0]
+### Added
+Added a preview for generated Code Fixes
```
@bastiandoetsch I wasnt sure how much in detail to go here. Is this summary enough?

bastiandoetsch replied:
That's ok. I would maybe rephrase to "Improve UX of AI fixes by adding previews and options". But it's also ok as is. The version number though does not seem correct. I'd assume 2.6.0 would be the release version (currently, we're at 2.5.0).

Arvi3d commented:
👏 👏 👏
